cmd/fix-usernames/main.go

```diff
@@ -0,0 +1,317 @@
+		Order("usages.last_modified_at DESC").
+		Limit(1).
+		Find(&usages).
+		Error
+
+	var usage *model.Usage
+	if len(usages) > 0 {
+		usage = &usages[0]
+	}
+	return usage, err
+}
+
+// findQuotaValue finds the quota value for a specific resource type in a list of quotas.
+func findQuotaValue(quotas []model.Quota, resourceTypeName string) float64 {
+	var quotaValue float64
+	for _, quota := range quotas {
+		if quota.ResourceType.Name == resourceTypeName && quota.Quota > quotaValue {
+			quotaValue = quota.Quota
+		}
+	}
+	return quotaValue
+}
+
+// setQuota either adds a quota to a subscription or updates a quota in a subscription.
+func setQuota(ctx context.Context, tx *gorm.DB, subscriptionID, resourceTypeID *string, quotaValue float64) error {
+	quota := model.Quota{
+		UserPlanID:     subscriptionID,
+		ResourceTypeID: resourceTypeID,
+		Quota:          quotaValue,
+	}
+	err := tx.WithContext(ctx).
+		Clauses(clause.OnConflict{
+			Columns: []clause.Column{
+				{
+					Name: "user_plan_id",
+				},
+				{
+					Name: "resource_type_id",
+				},
+			},
+			DoUpdates: clause.AssignmentColumns([]string{"quota"}),
+		}).
+		Create(&quota).Error
+	return err
+}
+
+// restorePreviousQuotas ensures that the resource usage limits for the new subscription are at least as large as the
+// resource usage limits for the old subscription.
+func restorePreviousQuotas(ctx context.Context, tx *gorm.DB, oldSubscription, newSubscription *model.UserPlan) error {
+	for _, quota := range oldSubscription.Quotas {
+		newQuotaValue := findQuotaValue(newSubscription.Quotas, quota.ResourceType.Name)
+		if newQuotaValue < quota.Quota {
+			err := setQuota(ctx, tx, newSubscription.ID, quota.ResourceType.ID, quota.Quota)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// addUsageToSubscription adds a usage record to the new subscription. This function is only intended to be used to
+// add usage records to a brand new subscription, so it assumes that there aren't any usages associated with the
+// subscription yet.
+func addUsageToSubscription(ctx context.Context, tx *gorm.DB, subscription *model.UserPlan, usage *model.Usage) error {
+	newUsage := &model.Usage{
+		ResourceTypeID: usage.ResourceTypeID,
+		UserPlanID:     subscription.ID,
+		Usage:          usage.Usage,
+	}
+	return tx.WithContext(ctx).Create(newUsage).Error
+}
+
+// fixUsername fixes a username for a single user.
+func fixUsername(ctx context.Context, tx *gorm.DB, oldUser *model.User, usernameSuffix string) error {
+	fmt.Printf("fixing the subscriptions for %s...\n", oldUser.Username)
+
+	// Get the information for the correct username.
+	newUsername := strings.TrimSuffix(oldUser.Username, usernameSuffix)
+	newUser, err := db.GetUser(ctx, tx, newUsername)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get the user details for %s", newUsername)
+	}
+
+	// Load the currently oldSubscription for the user.
+	oldSubscription, err := loadCurrentSubscription(ctx, tx, newUser)
```

Comment:
Shouldn't this be loading based on the `oldUser.Username`? I'd assume otherwise we wouldn't need to recreate the subscription or anything like that.

Reply:
I should have documented this more clearly; it looks like I was sleeptyping when I wrote that. 🤣 The reason that `newUser` is used in this case is because all existing subscriptions are being canceled, and new subscriptions with the same quotas are being created. Since subscription purchases never used the username suffix, using the username without the suffix ensures that we'll catch all subscriptions that were purchased. I'll fix the comment.